go/arrow/avro/schema.go

```diff
@@ -0,0 +1,409 @@
+			avroLogicalToArrowField(n)
+		} else {
+			n.arrowField = arrow.Field{Name: n.Name(), Type: &arrow.FixedSizeBinaryType{ByteWidth: n.schema.(*avro.FixedSchema).Size()}, Nullable: true}
+		}
+	case "string":
+		if isLogicalSchemaType(n.schema) {
+			avroLogicalToArrowField(n)
+		} else {
+			n.arrowField = arrow.Field{Name: n.name, Type: avroPrimitiveToArrowType(string(st)), Nullable: true}
+		}
+	case "bytes":
+		if isLogicalSchemaType(n.schema) {
+			avroLogicalToArrowField(n)
+		} else {
+			n.arrowField = arrow.Field{Name: n.name, Type: avroPrimitiveToArrowType(string(st)), Nullable: true}
+		}
+	case "int":
+		if isLogicalSchemaType(n.schema) {
+			avroLogicalToArrowField(n)
+		} else {
+			n.arrowField = arrow.Field{Name: n.name, Type: avroPrimitiveToArrowType(string(st)), Nullable: true}
+		}
+	case "long":
+		if isLogicalSchemaType(n.schema) {
+			avroLogicalToArrowField(n)
+		} else {
+			n.arrowField = arrow.Field{Name: n.name, Type: avroPrimitiveToArrowType(string(st)), Nullable: true}
+		}
+	case "float":
+		n.arrowField = arrow.Field{Name: n.name, Type: avroPrimitiveToArrowType(string(st)), Nullable: true}
+	case "double":
+		n.arrowField = arrow.Field{Name: n.name, Type: avroPrimitiveToArrowType(string(st)), Nullable: true}
+	case "boolean":
+		n.arrowField = arrow.Field{Name: n.name, Type: avroPrimitiveToArrowType(string(st)), Nullable: true}
+	case "<ref>":
+		n.schema = n.schemaCache.Get(n.schema.(*avro.RefSchema).Schema().Name())
+		arrowSchemafromAvro(n)
+	case "null":
+		n.arrowField = arrow.Field{Name: n.name, Type: arrow.Null, Nullable: true}
+	}
+}
+
+// iterate record Fields()
+func iterateFields(n *schemaNode) {
+	for _, f := range n.schema.(avro.PropertySchema).(*avro.RecordSchema).Fields() {
+		switch ft := f.Type().(type) {
+		// Avro "array" field type
+		case *avro.ArraySchema:
+			n.schemaCache.Add(f.Name(), ft.Items())
+			// logical items type
+			c := n.newChild(f.Name(), ft.Items())
+			if isLogicalSchemaType(ft.Items()) {
+				avroLogicalToArrowField(c)
+			} else {
+				arrowSchemafromAvro(c)
+			}
+			c.arrowField = arrow.Field{Name: c.name, Type: arrow.ListOf(c.arrowField.Type), Metadata: c.arrowField.Metadata, Nullable: true}
+		// Avro "enum" field type = Arrow dictionary type
+		case *avro.EnumSchema:
+			n.schemaCache.Add(f.Name(), f.Type())
+			c := n.newChild(f.Name(), f.Type())
+			symbols := make(map[string]string)
+			for index, symbol := range ft.Symbols() {
+				k := strconv.FormatInt(int64(index), 10)
+				symbols[k] = symbol
+			}
+			var dt arrow.DictionaryType = arrow.DictionaryType{IndexType: arrow.PrimitiveTypes.Uint64, ValueType: arrow.BinaryTypes.String, Ordered: false}
+			sl := len(symbols)
+			switch {
+			case sl <= math.MaxUint8:
+				dt.IndexType = arrow.PrimitiveTypes.Uint8
+			case sl > math.MaxUint8 && sl <= math.MaxUint16:
+				dt.IndexType = arrow.PrimitiveTypes.Uint16
+			case sl > math.MaxUint16 && sl <= math.MaxUint32:
+				dt.IndexType = arrow.PrimitiveTypes.Uint32
+			}
```

Comment:
I'll cast it to int64 with ```sl := int64(len(symbols))``` in both spots.
IIRC the duplication is necessary because one function handles non-nested fields and the other handles nested fields.